waspc/src/StrongPath.hs

```diff
@@ -0,0 +1,55 @@
+module StrongPath
+    ( Path
+    , Abs, File, Dir
+    , (</>)
+    , fromRelDir, fromRelFile, fromAbsDir, fromAbsFile
+    , toRelDir, toRelFile, toAbsDir, toAbsFile
+    ) where
+
+import qualified Path as P
+
+data Path a b = RelDir  (P.Path P.Rel P.Dir)
+              | RelFile (P.Path P.Rel P.File)
+              | AbsDir  (P.Path P.Abs P.Dir)
+              | AbsFile (P.Path P.Abs P.File)
+
+data Abs
+data File
+data Dir a
+
+(</>) :: Path a b -> Path b c -> Path a c
+(RelDir p1) </> (RelFile p2) = RelFile $ p1 P.</> p2
+(RelDir p1) </> (RelDir p2) = RelDir $ p1 P.</> p2
+(AbsDir p1) </> (RelFile p2) = AbsFile $ p1 P.</> p2
+(AbsDir p1) </> (RelDir p2) = AbsDir $ p1 P.</> p2
+_ </> _ = error "This can't theoretically happen."
+
+
+fromRelDir :: P.Path P.Rel P.Dir -> Path (Dir a) (Dir b)
+fromRelDir p = RelDir p
+
+fromRelFile :: P.Path P.Rel P.File -> Path (Dir a) File
+fromRelFile p = RelFile p
+
+fromAbsDir :: P.Path P.Abs P.Dir -> Path Abs (Dir a)
+fromAbsDir p = AbsDir p
+
+fromAbsFile :: P.Path P.Abs P.File -> Path Abs File
+fromAbsFile p = AbsFile p
+
+
+toRelDir :: Path (Dir a) (Dir b) -> P.Path P.Rel P.Dir
+toRelDir (RelDir p) = p
+toRelDir _ = error "This can't theoretically happen."
```

Comment:
maybe extract this error msg or the whole function somewhere, so you don't have to keep hardcoding error msg everywhere?

Reply:
Ok done!